src/lib.rs

```diff
@@ -182,93 +183,16 @@ pub struct SystemHandle {
-    pub fn send_once_to_self(&mut self, delay: Duration, msg: A::Message) -> ScheduleToken {
-        self.send_once_to::<A>(delay, msg, self.myself.recipient.clone())
-    }
-
-    /// Sends a message once to a specified actor recipient, after a delay.
-    /// Cancel with a call to `context.cancel_timer(schedule_token);`
-    pub fn send_once_to<R>(
-        &mut self,
-        delay: Duration,
-        msg: R::Message,
-        recipient: Recipient<R::Message>,
-    ) -> ScheduleToken
-    where
-        R: Actor + ?Sized,
-    {
-        self.timer.run_once(delay, move |_| {
-            if let Err(e) = recipient.send(msg) {
-                warn!("Error in send_once_to: {}", e);
-            }
-        })
-    }
-
-    /// Sends a recurring message to the calling actor itself, after a delay.
-    /// After the delay, messages will continue to be delivered every `interval` duration.
-    /// Ideally `msg_producer` should simply create a message and that's it. Do not do any
-    /// blocking work or you will block the timer thread.
-    /// Cancel with a call to `context.cancel_timer(schedule_token);`
-    pub fn send_recurring_to_self<F>(
-        &mut self,
-        delay: Duration,
-        interval: Duration,
-        msg_producer: F,
-    ) where
-        F: FnMut() -> A::Message + Send + 'static,
-    {
-        self.send_recurring_to::<F, A>(
-            delay,
-            interval,
-            msg_producer,
-            self.myself.recipient.clone(),
-        );
-    }
-
-    /// Sends a recurring message to a specified actor recipient, after a delay.
-    /// After the delay, messages will continue to be delivered every `interval` duration.
-    /// Ideally `msg_producer` should simply create a message and that's it. Do not do any
-    /// blocking work or you will block the timer thread.
-    /// Cancel with a call to `context.cancel_timer(schedule_token);`
-    pub fn send_recurring_to<F, R>(
-        &mut self,
-        delay: Duration,
-        interval: Duration,
-        mut msg_producer: F,
-        recipient: Recipient<R::Message>,
-    ) -> ScheduleToken
-    where
-        F: FnMut() -> R::Message + Send + 'static,
-        R: Actor + ?Sized,
-    {
-        self.timer.run_recurring(delay, interval, move |_| {
-            if let Err(e) = recipient.send(msg_producer()) {
-                warn!("Error in send_recurring_to: {}", e);
-            }
-        })
-    }
-
-    /// Cancel any timer-based message sending which was initiated via `send_*`.
-    /// The `ScheduleToken` comes from the various `send_*` calls.
-    /// Do not rely on precise timing of the cancellation, the message _may_ be sent
-    /// before it is cancelled.
-    pub fn cancel_timer(&mut self, schedule_token: ScheduleToken) {
-        self.timer.cancel(schedule_token);
-    }
-
-    pub fn shutdown(&mut self) -> Result<(), ActorError> {
-        self.system_handle.shutdown(&mut self.timer)
-    }
+    /// A deadline for receiving the next message.
```

Comment:
My old C++ self tells me that `receive_deadline` should be private, and there should be `set_deadline()`. Is `Rust` not opinionated about it?

Reply:
I was hesitant whether to make it pub or use getters/setters. Rust resources aren't unanimous on this:
- https://rust-lang.github.io/api-guidelines/future-proofing.html#c-struct-private warns that pub fields are a stronger commitment, and goes on to say:
  > Public fields are most appropriate for struct types in the C spirit: compound, passive data structures. Otherwise, consider providing getter/setter methods and hiding fields instead.
- https://www.reddit.com/r/rust/comments/d7w6n7/is_it_idiomatic_to_write_setters_and_getters/f15ib88?utm_source=share&utm_medium=web2x&context=3 leans on the "pub fields" side, noting rust specifics to destructuring, but is somewhat opinionated.
- https://users.rust-lang.org/t/public-getter-method-vs-pub-field/20147 contains very informative unopinionated summary of all implications, highlighting construction and partial borrowing. I liked matklad's (the author of rust-analyzer) rule of thumb there:
  > - If there’s a combination of values of fields which breaks some invariant, than make all fields private and provide a getter
  > - If any combination of fields values makes sense, make all fields public
- there were even bug reports to rust projects to make private fields `pub`, addressed: https://github.com/rust-syndication/rss/issues/71

In our case, all permissible values of `receive_deadline: Option<Instant>` are valid at any point of time, and it doesn't interact with other fields. That, and simplicity probably made me lean to the "field side".

Adding a private field would have a side-effect: it would make whole `Context` non-constructible by external crates unless we provided a `pub new(...) method. That would interact with actor wrapping unless we added some `pub wrap(...) method. Independently of PR, I actually think that might be a good idea, but it would probably fit better as a separate PR (and if done using `#[non_exhaustive]` we could keep the fields public if we wish so).

Reply:
My gut feeling is that we want this private, with some convenience functions for setting a specific deadline (`context.set_deadline(Instant)`) or timing out after some duration (`context.expire_deadline_after(Duration)`)

Reply:
Alright, I've changed this to private and introduced `deadline()`, `set_deadline()` and `set_timeout()` methods in a just-pushed version. I'll deal with the wrap-ability of `Context` in #54 later.

Reply:
Thank you for all the discussion pointers @strohel. To `matklad`'s point, `Context` is not a passive data container, and it's easy to see that it may evolve in the future to break the invariance. As you suggested, I think it makes sense in a follow-up PR to make all the fields private. It should by by design that the clients are not able to construct `Context`, and it's only given by the `actor` system.